### DISCUSSION TITLE
use of ESPHost library as a transport layer under WiFiNINA library

### DISCUSSION THREAD

**Discussion by lyusupov** [Networking-for-Arduino/ESPHost#3] [Ideas]
Hello Juraj!

Thank you very much for development of  the [ESPHost](https://github.com/JAndrassy/ESPHost) library!

I appreciate your efforts to transfer the technology from Arduino Renesas Core and making it available for RP2040's Core developed by Earle F. Philhower.
I was able to test it with 3.7.x RP2040 Cores and can confirm that it works nicely for me!

Juraj, you and  Earle  have created a medium layer in **ESP8266WiFi library** _(in use by RP2040 Core)_ to operate on top of ESPHost library. I can see that you have also created this layer for Mbed: https://github.com/JAndrassy/ESPHost-EMAC
ESP8266WiFi and Mbed are popular ones in the Arduino world, but IMHO the WiFiNINA library is _(currently)_ a bit more widely used.

I have a suggestion to create **a similar layer but in WiFiNINA library**, so the ESP-Host technology and your ESPHost library could operate with nRF52, SAMD, EFR32 and other Arduino platforms as well.

How do you estimate chances of this feature to become available in the future ?

**Comment by JAndrassy** [comment]
it is not possible to do it with WiFiNINA library. ESPHost is only a driver. The TCP/IP part is done by LwIP. Mbed Core and Pico Core have LwIP so it was enough to provide a driver.  The other cores don't have LwIP. For WiFiNINA library LwIP is in the ESP32.

I have a plan to create a LwIP base Arduino library, but I have that plan a few years now. 
LwIP requires a lot of RAM. For example it will not work on Arduino MKR or Zero (SAMD21), but It will work on SAMD51 (Adafruit M4 boards).

**Comment by lyusupov** [comment]
Thank you for quick response!

It explains a lot! We can see that even within one Arduino Renesas Core they two use different approaches:
1) ESP-Host for [PORTENTA_C33](https://github.com/arduino/ArduinoCore-renesas/tree/main/variants/PORTENTA_C33) with sufficient RAM available and
2) WiFiNINA-like "bridge" approach for low memory Uno R4 WiFi (32K RAM in RA4M1)

I have one question remaining about nRF52840.
Just like RP2040, the nRF52840 has sufficient memory to implement lwIP stack.
However, the [Adafruit Core](https://github.com/adafruit/Adafruit_nRF52_Arduino) for nRF52 does not contain any of IP features built-in.

If we will decide to transfer ESP8266WiFi library approach from Earle's RP2040 to Adafruit's nRF52 Core, what do you think can cause substantial issues to solve?
As an example, I can see that  Earle's RP2040 Core is using **lwIP stack located inside Pico-SDK**. However, it does not look that [Nordic SDK](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfx/api_reference.html) _( in use by Adafruit's nRF52 Core )_ contains any of lwIP implementation available. Thus, this could be an issue, right ?

**Comment by JAndrassy** [comment]
I made the ESPHost work with Uno R4, but the RAM was almost full. then they rebuild the LwIP for better performance and it couldn't fit into Uno R4 RAM anymore.

Adafruit doesn't focus on networking. If I make the lwip based library it will support nRF52 too. I didn't say in previous comment that the library would support Ethernet at least with W5500 and ENC28J60 and WiFi at least with ESPHost and ATWINC1500 (WiFi101 hw). (currently I am slowly working on the WINC1500 driver for Pico Core)

There is the QNEthernet lwip based library. It was written for the Teensy 4.1 Ethernet peripheral, but I [inspired](https://github.com/ssilverman/QNEthernet/discussions/62) the author to port it to STM32 and maybe later to other platforms with Wiznet W5500 driver for now. I don't know if he would want to enhance the library for WiFi.
